### PR TITLE
PERF: IOV count support added

### PR DESCRIPTION
**Pull request by shssf**
Implemented IOV list options in UCT part of the test
This is related to #951

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2977/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/495/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2442/ for details (Mellanox internal link).

File: src/tools/perf/libperf.c [hunk 1]
```diff
@@ -120,6 +122,16 @@ static ucs_status_t uct_perf_test_alloc_mem(ucx_perf_context_t *perf,
     ucs_assert(perf->uct.recv_mem.md == perf->uct.md);
     perf->recv_buffer = perf->uct.recv_mem.address;
 
+    /* Allocate IOV datatype memory */
+    perf->params.uct.iovcnt = params->uct.iovcnt ? params->uct.iovcnt : 1;
+    perf->uct.iov = malloc(sizeof(*perf->uct.iov) * perf->params.uct.iovcnt *
+                       params->thread_count);
```
Comment (yosefe) [inline src/tools/perf/libperf.c]: indentation

File: src/tools/perf/libperf.c [hunk 2]
```diff
@@ -428,6 +439,17 @@ static ucs_status_t uct_perf_test_check_capabilities(ucx_perf_params_t *params,
         }
     }
 
+    if ((UCT_PERF_DATA_LAYOUT_ZCOPY == params->uct.data_layout)) {
+        if (params->uct.iovcnt > max_iov) {
+            if (params->flags & UCX_PERF_TEST_FLAG_VERBOSE) {
+                ucs_error("Number of IOV entries exceeded. Maximum is -i %lu", max_iov);
```
Comment (yosefe) [inline src/tools/perf/libperf.c]: pls remove "-i" - it's only a paramter for perftest application, but libperf can be also used w/o it, for example from gtest

File: src/tools/perf/libperf.c [hunk 3]
```diff
@@ -428,6 +439,17 @@ static ucs_status_t uct_perf_test_check_capabilities(ucx_perf_params_t *params,
...
+            }
+            return UCS_ERR_UNSUPPORTED;
+        } else if (0 == params->uct.iovcnt) {
+            params->uct.iovcnt = max_iov;
```
Comment (yosefe) [inline src/tools/perf/libperf.c]: iovcnt==0 should also be an error

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2986/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/512/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/513/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2987/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/515/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2452/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2988/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2453/ for details (Mellanox internal link).

**Comment by shssf** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/525/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2459/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/529/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/531/ for details.

File: src/tools/perf/libperf_int.h [hunk 4]
```diff
@@ -136,6 +137,66 @@ static inline void ucx_perf_update(ucx_perf_context_t *perf, ucx_perf_counter_t
...
+            } else {
+                ucs_assert_always(perf->params.msg_size_lst[iov_it] >= header_size);
+                iov[iov_it].length = perf->params.msg_size_lst[iov_it] - header_size;
+            }
+            iov[iov_it].memh       = memh;
+            iov[iov_it].stride     = 0;
+            iov[iov_it].count      = 1;
+
+            if (perf->params.iov_stride) {
+                ucs_assert_always(iov[iov_it].length <= perf->params.iov_stride);
```
Comment (yosefe) [inline src/tools/perf/libperf_int.h]: this should be checked in earlier stage and produce "invalid parameter" error

File: src/tools/perf/libperf_int.h [hunk 5]
```diff
@@ -136,6 +137,66 @@ static inline void ucx_perf_update(ucx_perf_context_t *perf, ucx_perf_counter_t
...
+        ucs_debug("IOV buffer filled by %lu slices with total length %lu",
+                  iovcnt, iov_length_it);
+}
+
+/**
+ * Get the length between beginning of the IOV first buffer and the latest byte
+ * in the latest IOV buffer. Has meaning if stride > 0 and zcopy layout only.
+ */
+static inline
+size_t ucx_perf_get_control_shift(const ucx_perf_context_t *perf)
```
Comment (yosefe) [inline src/tools/perf/libperf_int.h]: rename, for example to to ucx_perf_get_buffer_extent

File: src/tools/perf/perftest.c [hunk 6]
```diff
@@ -324,11 +324,12 @@ static void usage(struct perftest_context *ctx, const char *program)
     printf("     -x <tl>        Transport to use for testing.\n");
     printf("     -c <cpu>       Set affinity to this CPU. (off)\n");
     printf("     -n <iters>     Number of iterations to run. (%ld)\n", ctx->params.max_iter);
-    printf("     -s <size>      Message size. (%zu)\n", ctx->params.message_size);
+    printf("     -s <size>      Message size. default is (%zu). It can contains (applicable in zcopy data layout only) a list of message sizes, separated by comma. For example, \"-s 16,48,8192,8192,14\"\n", ctx->params.message_size);
```
Comment (yosefe) [inline src/tools/perf/perftest.c]: 1. "List of buffer sizes separated by comma, which make up a single message. Default is (%zu). "
1. limit any single line of code to 80 chars.

File: src/tools/perf/perftest.c [hunk 7]
```diff
@@ -324,11 +324,12 @@ static void usage(struct perftest_context *ctx, const char *program)
...
     printf("     -H <size>      AM Header size. (%zu)\n", ctx->params.am_hdr_size);
     printf("     -w <iters>     Number of warm-up iterations. (%zu)\n", ctx->params.warmup_iter);
     printf("     -W <count>     Flow control window size, for active messages. (%u)\n", ctx->params.uct.fc_window);
     printf("     -O <count>     Maximal number of uncompleted outstanding sends. (%u)\n", ctx->params.max_outstanding);
+    printf("     -i <count>     Stride between beginnings of payload elements in the buffer in bytes (applicable in zcopy data layout only). The same as UCT uct_iov_t stride.\n");
```
Comment (yosefe) [inline src/tools/perf/perftest.c]: This is not for "zcopy" only, it would be used for UCP as well.
"Stride between consecutive buffers in the same message. 0 means the buffers are allocated contiguously. Otherwise, must not be smaller than any of the buffer sizes specified by -s"

File: src/tools/perf/perftest.c [hunk 8]
```diff
@@ -424,11 +431,45 @@ static ucs_status_t parse_test_params(ucx_perf_params_t *params, char opt, const
             return -1;
         }
         return UCS_OK;
+    case 'i':
+        params->iov_stride = atol(optarg);
+        if (!params->iov_stride) {
```
Comment (yosefe) [inline src/tools/perf/perftest.c]: 0 is also a valid value

File: src/tools/perf/perftest.c [hunk 9]
```diff
@@ -424,11 +431,45 @@ static ucs_status_t parse_test_params(ucx_perf_params_t *params, char opt, const
...
+        return UCS_OK;
     case 'n':
         params->max_iter = atol(optarg);
         return UCS_OK;
     case 's':
-        params->message_size = atol(optarg);
+        optarg_ptr = (char *)optarg;
+        token_num  = 0;
+        /* count the number of given message sizes */
+        while((optarg_ptr = strchr(optarg_ptr, delim))) {
```
Comment (yosefe) [inline src/tools/perf/perftest.c]: 1. space after while
1. pls remove duplicate (

File: src/tools/perf/perftest.c [hunk 10]
```diff
@@ -424,11 +431,45 @@ static ucs_status_t parse_test_params(ucx_perf_params_t *params, char opt, const
...
+            ++optarg_ptr;
+            ++token_num;
+        }
+        ++token_num;
+
+        params->msg_size_lst = malloc(sizeof(*params->msg_size_lst) * token_num);
```
Comment (yosefe) [inline src/tools/perf/perftest.c]: pls extract the code below (size list parsing) to another helper function

File: src/tools/perf/perftest.c [hunk 11]
```diff
@@ -1111,6 +1169,9 @@ static ucs_status_t run_test_recurs(struct perftest_context *ctx,
     while ((status = read_batch_file(batch_file, &params, &ctx->test_names[depth])) == UCS_OK) {
         status = run_test_recurs(ctx, &params, depth + 1);
         free(ctx->test_names[depth]);
+        if (NULL == parent_params->msg_size_lst && NULL != params.msg_size_lst) {
```
Comment (yosefe) [inline src/tools/perf/perftest.c]: pls add ( ) for each expression

File: src/tools/perf/perftest.c [hunk 12]
```diff
@@ -1111,6 +1169,9 @@ static ucs_status_t run_test_recurs(struct perftest_context *ctx,
...
+            free(params.msg_size_lst);
```
Comment (yosefe) [inline src/tools/perf/perftest.c]: who releases params.msg_size_lst in non-batch mode?

File: test/gtest/common/test_perf.cc [hunk 13]
```diff
@@ -196,6 +196,9 @@ test_perf::test_result test_perf::run_multi_threaded(const test_spec &test, unsi
     strncpy(params.uct.tl_name , tl_name.c_str(),  sizeof(params.uct.tl_name));
     params.uct.data_layout = test.data_layout;
     params.uct.fc_window   = UCT_PERF_TEST_MAX_FC_WINDOW;
+    params.msg_size_lst    = NULL;
```
Comment (yosefe) [inline test/gtest/common/test_perf.cc]: need to add some IOV tests to gtest as well

File: src/tools/perf/libperf.c [hunk 14]
```diff
@@ -428,6 +464,27 @@ static ucs_status_t uct_perf_test_check_capabilities(ucx_perf_params_t *params,
...
+        if (params->iovcnt > max_iov) {
+            if ((params->flags & UCX_PERF_TEST_FLAG_VERBOSE) || !params->iovcnt) {
+                ucs_error("Wrong number of IOV entries. Requested is %lu,"
+                          " should be in the range 1...%lu", params->iovcnt, max_iov);
+            }
+            return UCS_ERR_UNSUPPORTED;
+        }
+        /* if iovcnt == 1 the message size checked above */
+        if ((UCX_PERF_CMD_AM == params->command) && (params->iovcnt > 1)) {
+            if (params->am_hdr_size > params->msg_size_lst[0]) {
```
Comment (shssf) [inline src/tools/perf/libperf.c]: The AM header used the buffer allocated initially (not additional memory but inlined into the buffer). To convert this memory into IOV I need to check if the AM header can be fit into first IOV entry. The size of the first IOV entry will be `iov[0].length - am_header_size`.

File: src/tools/perf/perftest.c (hunk 9 above)
Comment (shssf) [inline src/tools/perf/perftest.c]: duplicate `()` just to avoid warning "possible assumption" at least in eclipse.

Comment (yosefe) [inline src/tools/perf/perftest.c]: interesting, IMHO this is better:
` while((optarg_ptr = strchr(optarg_ptr, delim)) != NULL) {`

File: test/gtest/common/test_perf.cc (hunk 13 above)
Comment (shssf) [inline test/gtest/common/test_perf.cc]: I think it would be better to implement it as separate PR. I created the issue #1186 for this. 
It might be good idea to label it as "training"

File: src/tools/perf/libperf.c (hunk 14 above)
Comment (yosefe) [inline src/tools/perf/libperf.c]: so why not add am_header_size to the size of the buffer?

File: test/gtest/common/test_perf.cc (hunk 13 above)
Comment (yosefe) [inline test/gtest/common/test_perf.cc]: i think this should be done here, at least 1-2 test cases. i'm afraid some issues such as memory leaks could be missed if this is not tested at all inside gtest.

File: src/tools/perf/libperf.c (hunk 14 above)
Comment (shssf) [inline src/tools/perf/libperf.c]: Of course we can. It is required changes in overall test logic related to this header. If you think it is important, let's make these changes in next PR.

Comment (yosefe) [inline src/tools/perf/libperf.c]: i see. so let's keep it as-is for now.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3026/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/567/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2507/ for details (Mellanox internal link).

File: src/tools/perf/libperf.h [hunk 15]
```diff
@@ -141,7 +141,13 @@ typedef struct ucx_perf_params {
     ucx_perf_wait_mode_t   wait_mode;       /* How to wait */
     unsigned               flags;           /* See ucx_perf_test_flags. */
 
-    size_t                 message_size;    /* Test message size */
+    size_t                 *msg_size_list;  /* Test message sizes list. The size
+                                               of the array is in iovcnt */
```
Comment (yosefe) [inline src/tools/perf/libperf.h]: s/iovcnt/msg_size_cnt/

File: src/tools/perf/perftest.c [hunk 16]
```diff
@@ -361,6 +374,48 @@ static const char *__basename(const char *path)
...
+    optarg_ptr = (char *)optarg;
+    token_num  = 0;
+    /* count the number of given message sizes */
+    while ((optarg_ptr = strchr(optarg_ptr, delim)) != NULL) {
+        ++optarg_ptr;
+        ++token_num;
+    }
+    ++token_num;
+
+    if (params->msg_size_list) { /* free previously allocated buffer */
```
Comment (yosefe) [inline src/tools/perf/perftest.c]: free(NULL) is allowed - no need for 'if'

File: src/tools/perf/perftest.c [hunk 17]
```diff
@@ -361,6 +374,48 @@ static const char *__basename(const char *path)
...
+    }
+    params->msg_size_list = malloc(sizeof(*params->msg_size_list) * token_num);
+    if (NULL == params->msg_size_list) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    optarg_ptr = (char *)optarg;
+    errno = 0;
+    for (token_it = 0; token_it < token_num; ++token_it) {
+        params->msg_size_list[token_it] = strtoul(optarg_ptr, &optarg_ptr2, 10);
```
Comment (yosefe) [inline src/tools/perf/perftest.c]: could pass base=0 to strtoul() , to doing "-s 0x100,0x200" would also work

File: test/gtest/uct/test_uct_perf.cc [hunk 18]
```diff
@@ -25,77 +25,77 @@ test_perf::test_spec test_uct_perf::tests[] =
...
   { "am bcopy bw", "MB/sec",
     UCX_PERF_API_UCT, UCX_PERF_CMD_AM, UCX_PERF_TEST_TYPE_STREAM_UNI,
-    UCT_PERF_DATA_LAYOUT_BCOPY, 1000, 1, 100000l,
+    UCT_PERF_DATA_LAYOUT_BCOPY, 0, 1, { 1000 }, 1, 100000l,
     ucs_offsetof(ucx_perf_result_t, bandwidth.total_average), MB, 620.0, 15000.0 },
 
   { "am zcopy bw", "MB/sec",
     UCX_PERF_API_UCT, UCX_PERF_CMD_AM, UCX_PERF_TEST_TYPE_STREAM_UNI,
-    UCT_PERF_DATA_LAYOUT_ZCOPY, 1000, 32, 100000l,
+    UCT_PERF_DATA_LAYOUT_ZCOPY, 8192, 3, { 256, 256, 512 }, 32, 100000l,
```
Comment (yosefe) [inline test/gtest/uct/test_uct_perf.cc]: would be better to create another test (e.g "am iov bw") instead of changing existing one, in case the transport would not support such amount of IOVs

**Comment by yosefe** [comment]
overall looks good, except few minor comments.
are you planning to add UCP/IOV support to perftest in another PR?

**Comment by shssf** [comment]
@yosefe  Yes, I planned PERF/UCP a bit later because it requires adding datatype support in it.
I will address your comments, in this PR, soon.

File: src/tools/perf/perftest.c (hunk 17 above)
Comment (shssf) [inline src/tools/perf/perftest.c]: @yosefe  I did this to avoid user confusion in case of `-s 0100,0200` (octal base).
Do we want this behaviour?

Comment (yosefe) [inline src/tools/perf/perftest.c]: ok, let's keep as-is

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3050/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/601/ for details.

**Comment by shssf** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/603/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2542/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3061/ for details.
